File: Leetcode/202. Happy Number/happyNumbers.rb
# https://leetcode.com/problems/happy-number/

def isHappy(n)
  def sumNumbers(num)
    total_sum = 0

    while num > 0
      digit = num % 10
      num = (num / 10).floor()
      total_sum += digit ** 2
    end

    total_sum
  end

  slow = sumNumbers(n)
  fast = sumNumbers(sumNumbers(n))

  while true
    if fast == 1
      return true
    elsif fast == slow
      return false
    else
      slow = sumNumbers(slow)
      fast = sumNumbers(sumNumbers(fast))
    end
  end
end

File: Leetcode/202. Happy Number/HappyNumbers.java
// https://leetcode.com/problems/happy-number/

public class HappyNumbers {

  public static int sumNumbers(int num) {
    int totalSum = 0;

    while (num > 0) {
      int digit = num % 10;
      num = (int) Math.floor(num / 10);
      totalSum += digit * digit;
    }

    return totalSum;
  }

  public static boolean isHappy(int n) {

    int slow = sumNumbers(n);
    int fast = sumNumbers(sumNumbers(n));

    while (true) {
      System.out.println("slow es : " + slow);
      System.out.println("fast es: " + fast);
      if (fast == 1) {
        return true;
      } else if (fast == slow) {
        return false;
      } else {
        slow = sumNumbers(slow);
        fast = sumNumbers(sumNumbers(fast));
      }
    }
  }

  public static void main(String[] args) {
    System.out.println(isHappy(19));
  }
}
